library/mps/reader.c

```diff
@@ -0,0 +1,626 @@
+ *           remain intact even in the face of failing
+ *           and/or non-sensical calls.
+ *     Moreover, it might even reduce code-size because
+ *     the compiler need not write back temporary results
+ *     to memory in case of failure.
+ *
+ */
+
+int mbedtls_reader_init( mbedtls_reader *rd,
+                         unsigned char *acc, size_t acc_len )
+{
+    mbedtls_reader const zero = { NULL, 0, 0, 0, 0, NULL, 0, 0, { 0 } };
+    *rd = zero;
+    TRACE_INIT( "reader_init, acc len %u", (unsigned) acc_len );
+
+    rd->acc = acc;
+    rd->acc_len = acc_len;
+    RETURN( 0 );
+}
+
+int mbedtls_reader_free( mbedtls_reader *rd )
+{
+    mbedtls_reader const zero = { NULL, 0, 0, 0, 0, NULL, 0, 0, { 0 } };
+    TRACE_INIT( "reader_free" );
+    *rd = zero;
+    RETURN( 0 );
+}
+
+int mbedtls_reader_feed( mbedtls_reader *rd, unsigned char *new_frag,
+                         size_t new_frag_len )
+{
+    unsigned char *frag, *acc;
+    size_t copy_to_acc;
+    TRACE_INIT( "reader_feed, frag %p, len %u",
+                new_frag, (unsigned) new_frag_len );
+
+    if( new_frag == NULL )
+        RETURN( MBEDTLS_ERR_READER_INVALID_ARG );
+
+    for( size_t idx=0; idx < new_frag_len; idx++ )
+        TRACE( trace_comment, "Data[%u] = %02x", (unsigned) idx, (unsigned) new_frag[idx] );
+
+    /* Feeding is only possible if no fragment is currently being processed. */
+    frag = rd->frag;
+    if( frag != NULL )
+        RETURN( MBEDTLS_ERR_READER_UNEXPECTED_OPERATION );
+
+    acc = rd->acc;
+    if( acc != NULL )
+    {
+        size_t aa, ar;
+
+        ar = rd->acc_share.acc_remaining;
+        aa = rd->acc_avail;
+
+        copy_to_acc = ar;
+        if( copy_to_acc > new_frag_len )
+            copy_to_acc = new_frag_len;
+
+        acc += aa;
+        memcpy( acc, new_frag, copy_to_acc );
+
+        TRACE( trace_comment, "Copy new data of size %u of %u into accumulator at offset %u",
+                (unsigned) copy_to_acc, (unsigned) new_frag_len, (unsigned) aa );
+        for( size_t idx=0; idx < copy_to_acc; idx++ )
+        {
+            TRACE( trace_comment, "copy[%u] = %u", (unsigned) idx,
+                   (unsigned) new_frag[idx] );
+        }
+
+        /* Check if, with the new fragment, we have enough data. */
+        ar -= copy_to_acc;
+        if( ar > 0 )
+        {
+            /* Need more data */
+            aa += copy_to_acc;
+            rd->acc_share.acc_remaining = ar;
+            rd->acc_avail = aa;
+            RETURN( MBEDTLS_ERR_READER_NEED_MORE );
+        }
+
+        TRACE( trace_comment, "Enough data available to serve user request" );
+
+        rd->acc_share.frag_offset = aa;
+        aa += copy_to_acc;
+        rd->acc_avail = aa;
+    }
+    else
+    {
+        rd->acc_share.frag_offset = 0;
+    }
+
+    rd->frag = new_frag;
+    rd->frag_len = new_frag_len;
+    rd->commit = 0;
+    rd->end = 0;
+    RETURN( 0 );
+}
+
+
+int mbedtls_reader_get( mbedtls_reader *rd, size_t desired,
```

Comment:
I'd suggest refactoring the reader_get to separate two functions - serving the request from the current fragment, and separately, from the accumulator. These two cases only share a few lines of the variable preparation.